File: model-client/src/main/java/org/modelix/model/api/WriteTransaction.java
package org.modelix.model.api;

import org.jetbrains.annotations.NotNull;

public class WriteTransaction extends Transaction implements IWriteTransaction {
  protected ITree tree;
  private boolean closed = false;
  protected IIdGenerator idGenerator;

  public WriteTransaction(@NotNull ITree tree, IBranch branch, IIdGenerator idGenerator) {
    super(branch);
    this.tree = tree;
    this.idGenerator = idGenerator;
  }

  public void close() {
    closed = true;
  }

  protected void checkNotClosed() {
    if (closed) {
      throw new IllegalStateException("Transaction is already closed");
    }
  }

  @Override
  public ITree getTree() {
    return tree;
  }

  @Override
  public void setTree(ITree newTree) {
    checkNotClosed();
    this.tree = newTree;
  }

  @Override
  public void setProperty(long nodeId, String role, String value) {
    checkNotClosed();
    tree = getTree().setProperty(nodeId, role, value);
  }

  @Override
  public void setReferenceTarget(long sourceId, String role, INodeReference target) {
    checkNotClosed();
    tree = getTree().setReferenceTarget(sourceId, role, target);
  }

  @Override
  public void moveChild(long newParentId, String newRole, int newIndex, long childId) {
    checkNotClosed();
    tree = getTree().moveChild(newParentId, newRole, newIndex, childId);
  }

  @Override
  public long addNewChild(long parentId, String role, int index, IConcept concept) {
    checkNotClosed();
    long newId = idGenerator.generate();
    addNewChild(parentId, role, index, newId, concept);
    return newId;
  }

  @Override
  public void addNewChild(long parentId, String role, int index, long childId, IConcept concept) {
    checkNotClosed();
    tree = getTree().addNewChild(parentId, role, index, childId, concept);
  }

  @Override
  public void deleteNode(long nodeId) {
    checkNotClosed();
    getTree().getAllChildren(nodeId).forEach(this::deleteNode);
    tree = getTree().deleteNode(nodeId);
  }
}

File: model-client/src/main/java/org/modelix/model/api/WriteTransaction.kt
package org.modelix.model.api

class WriteTransaction(_tree: ITree?, branch: IBranch?, idGenerator: IIdGenerator) : Transaction(branch), IWriteTransaction {
    override var tree: ITree? = _tree
        set(value) {
            checkNotClosed()
            tree = value
        }
    private var closed = false
    protected var idGenerator: IIdGenerator
    fun close() {
        closed = true
    }

    protected fun checkNotClosed() {
        check(!closed) { "Transaction is already closed" }
    }

//    override fun getTree(): ITree? {
//        return tree
//    }
//
//    override fun setTree(newTree: ITree?) {
//        checkNotClosed()
//        tree = newTree
//    }

    override fun setProperty(nodeId: Long, role: String?, value: String?) {
        checkNotClosed()
        tree = tree!!.setProperty(nodeId, role, value)
    }

    override fun setReferenceTarget(sourceId: Long, role: String?, target: INodeReference?) {
        checkNotClosed()
        tree = tree!!.setReferenceTarget(sourceId, role, target)
    }

    override fun moveChild(newParentId: Long, newRole: String?, newIndex: Int, childId: Long) {
        checkNotClosed()
        tree = tree!!.moveChild(newParentId, newRole, newIndex, childId)
    }

    override fun addNewChild(parentId: Long, role: String?, index: Int, concept: IConcept?): Long {
        checkNotClosed()
        val newId = idGenerator.generate()
        addNewChild(parentId, role, index, newId, concept)
        return newId
    }

    override fun addNewChild(parentId: Long, role: String?, index: Int, childId: Long, concept: IConcept?) {
        checkNotClosed()
        tree = tree!!.addNewChild(parentId, role, index, childId, concept)
    }

    override fun deleteNode(nodeId: Long) {
        checkNotClosed()
        tree!!.getAllChildren(nodeId)!!.forEach { nodeId: Long -> deleteNode(nodeId) }
        tree = tree!!.deleteNode(nodeId)
    }

    init {
        this.tree = tree
        this.idGenerator = idGenerator
    }
}